### PR TITLE
fix: Vo application notifications are sent only to valid admins

### PR DESCRIPTION
**Pull request by Kuliak**
* added functionality to check whether each admin is a valid member of given Vo

### CODE REVIEW

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java [hunk 1]
```diff
@@ -1617,6 +1617,20 @@ public boolean haveStatus(PerunSession sess, Member member, Status status) {
 		return member.getStatus().equals(status);
 	}
 
+	@Override
+	public boolean isUserValidMember(PerunSession sess, User user, Vo vo) {
+		// check if given user is VALID member of VO
```
Comment (Johaney-s) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java:1622]: Looks good, just cosmetic suggestion - I'd say both comments are not necessary and can be removed.

**Review by Johaney-s — APPROVED** [review]

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/MailManagerImpl.java [hunk 2]
```diff
@@ -884,13 +884,7 @@ private void setFromAndReplyTo(MimeMessage message, String fromName, String from
...
 				// try to use all group admins with specified preferred emails
 				List<RichUser> admins = AuthzResolverBlImpl.getRichAdmins(registrarSession, app.getGroup(), Role.GROUPADMIN);
-				for (User admin : admins) {
-					attrSenderEmail = attrManager.getAttribute(registrarSession, admin, URN_USER_PREFERRED_MAIL);
-
-					if (attrSenderEmail != null && attrSenderEmail.getValue() != null) {
-						result.add(attrSenderEmail.valueAsString());
-					}
-				}
+				result.addAll(getUserPreferredMailsForValidMembers(registrarSession, admins, app.getVo()));
```
Comment (cuadradek) [inline perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/MailManagerImpl.java:887]: Groups can have admins assigned through a group from different VO. Therefore, we should check if they are valid in that VO, not in the application's VO. We should also check their status in the group through which they are assigned.

**Review by cuadradek — CHANGES_REQUESTED** [review]

File: perun-core/src/main/java/cz/metacentrum/perun/core/implApi/AuthzResolverImplApi.java [hunk 3]
```diff
@@ -552,7 +552,7 @@
 	void unsetRole(PerunSession sess, Map<String, Integer> mappingOfValues, String role) throws RoleNotSetException;
 
 	/**
-	 * Get all richUser administrators for complementary object and role with specified attributes.
+	 * Get all valid richUser administrators (both in VO and given group) for complementary object and role with specified attributes.
```
Comment (Johaney-s) [inline perun-core/src/main/java/cz/metacentrum/perun/core/implApi/AuthzResolverImplApi.java:555]: This comment should be added to all methods relying on this method.

**Review by Johaney-s — CHANGES_REQUESTED** [review]

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/MailManagerImpl.java [hunk 4]
```diff
@@ -2171,20 +2165,36 @@ private MimeMessage getInvitationMessage(Vo vo, Group group, String language, St
...
+	/**
+	 * Get preferred email for each user.
+	 *
+	 * @param sess
+	 * @param users
+	 * @return list of preferred emails for given users.
+	 * @throws AttributeNotExistsException
+	 * @throws WrongAttributeAssignmentException
+	 */
+	private List<String> getUserPreferredMailsForMembers(PerunSession sess, List<RichUser> users)
```
Comment (Johaney-s) [inline perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/MailManagerImpl.java:2188]:
```suggestion
	private List<String> getUserPreferredMails(PerunSession sess, List<RichUser> users)
```

**Review by Johaney-s — APPROVED** [review]

**Review by cuadradek — APPROVED** [review]

**Pull request merged**